R/cal-apply.R

```diff
@@ -0,0 +1,98 @@
+#---------------------------------- Apply --------------------------------------
+
+#' Applies a calibration to a set of prediction probabilities
+#' @details It currently supports data.frames only. It extracts the `truth` and
+#' the estimate columns names, and levels, from the calibration object.
+#' @param .data An object that can process a calibration object.
+#' @param calibration The calibration object (`cal_object`).
+#' @param ... Optional arguments; currently unused.
+#' @examples
+#' w_calibration <- cal_estimate_logistic(segment_logistic, Class)
+#'
+#' cal_apply(segment_logistic, w_calibration)
+#' @export
+cal_apply <- function(.data, calibration, ...) {
+  UseMethod("cal_apply")
+}
+
+#' @export
+cal_apply.data.frame <- function(.data, calibration, ...) {
+  if (calibration$type == "binary") {
+    cal_add_adjust(calibration, .data)
+  } else {
+    stop_multiclass()
+  }
+}
+
+# ------------------------------- Adjust ---------------------------------------
+
+cal_add_adjust <- function(calibration, .data) {
+  UseMethod("cal_add_adjust")
+}
+
+cal_add_adjust.cal_estimate_logistic <- function(calibration, .data) {
+  cal_add_predict_impl(
+    calibration = calibration,
+    .data = .data
+  )
+}
+
+cal_add_adjust.cal_estimate_logistic_spline <- function(calibration, .data) {
+  cal_add_predict_impl(
+    calibration = calibration,
+    .data = .data
+  )
+}
+
+cal_add_adjust.cal_estimate_isotonic_boot <- function(calibration, .data) {
+  cal_add_interval_impl(
+    calibration = calibration,
+    .data = .data
+  )
+}
+
+cal_add_adjust.cal_estimate_isotonic <- function(calibration, .data) {
+  cal_add_interval_impl(
+    calibration = calibration,
+    .data = .data
+  )
+}
+
+#---------------------------- Adjust implementations ---------------------------
+
+cal_add_predict_impl <- function(calibration, .data) {
+  if (calibration$type == "binary") {
+    model <- calibration$estimates
+    preds <- predict(model, newdata = .data, type = "response")
+    preds <- 1 - preds
+    .data[calibration$levels[[1]]] <- preds
```

Comment:
Why are the levels in a list? A named vector would be simpler.

Reply:
It is named, but the names are the factorial values. For binary models, it was easier just to always use the first level because that will be always the one the model ran against

Reply:
Ok, so why a list if it is a named character vector of factor levels?

Reply:
Because they are symbols and not the character factor levels